ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,592 @@
+        {
+            //Stop once whitespace is encountered
+            if (whiteSpace.Contains(c))
+            {
+                stateStack.Pop();
+                stateStack.Push(State.SEEK_VALUE);
+                currentName = currentString.ToString();
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read a structure
+        private static void ReadValueStruct(char c)
+        {
+            //Ignore whitespace
+            if (whiteSpace.Contains(c))
+            {
+                return;
+            }
+
+            //Catch comments
+            if (c == '/')
+            {
+                stateStack.Push(State.COMMENT);
+                currentString.Clear();
+                currentString.Append(c);
+                return;
+            }
+
+            //Check for the end of the structure
+            if (c == '}')
+            {
+                KVObject value = objStack.Pop();
+                objStack.Peek().AddProperty(value.key, new KVValue(typeof(KVObject), value));
+                stateStack.Pop();
+                return;
+            }
+
+            //Start looking for the next property name
+            stateStack.Push(State.PROP_NAME);
+            currentString.Clear();
+            currentString.Append(c);
+        }
+        
+        //Read a string value
+        private static void ReadValueString(char c)
+        {
+            if (c == '"' && fileStream[index-1] != '\\')
+            {
+                //String ending found
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(String), currentString.ToString()));
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Reading multiline string
+        private static void ReadValueStringMulti(char c)
+        {
+            //Check for ending
+            if (c == '"' && c == '"' && fileStream[index + 1] == '"' && fileStream[index + 2] == '"' && fileStream[index-1] != '\\')
+            {
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(String), currentString.ToString()));
+
+                //Skip to end
+                index += 2;
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read a boolean variable
+        private static void ReadValueBool(char c)
+        {
+            //Stop reading once the end of true or false is reached
+            if (c == 'e')
+            {
+                currentString.Append(c);
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(Boolean), currentString.ToString() == "true" ? true : false));
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read a numerical value
+        private static void ReadValueNumber(char c)
+        {
```

Comment:
Does this numeric implementation allow `345-678`, since '-' is in the set of numeric characters you've defined? I assume that would crash your parser? '-' should probably be only allowed if it's leading, as a unary negation operator.